src/common/base/src/base/dma.rs

```diff
@@ -0,0 +1,437 @@
+        let buf = buf.as_mut().unwrap();
+        unsafe { buf.set_len(buf.capacity()) };
+        match rustix::io::read(fd, buf) {
+            Ok(n) => {
+                unsafe { buf.set_len(n) };
+                Ok(n)
+            }
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    fn truncate(&self, length: usize) -> io::Result<()> {
+        rustix::fs::ftruncate(&self.fd, length as u64).map_err(|e| e.into())
+    }
+
+    async fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> {
+        self.fd.seek(pos).await
+    }
+}
+
+pub fn align_up(alignment: usize, value: usize) -> usize {
+    (value + alignment - 1) & !(alignment - 1)
+}
+
+pub fn align_down(alignment: usize, value: usize) -> usize {
+    value & !(alignment - 1)
+}
+
+async fn open_dma(file: File) -> io::Result<DmaFile> {
+    let stat = fstatvfs(&file).await?;
+    let alignment = stat.f_bsize.max(512) as usize;
+
+    Ok(DmaFile {
+        fd: file,
+        alignment,
+        buf: None,
+    })
+}
+
+async fn fstatvfs(file: &File) -> io::Result<rustix::fs::StatVfs> {
+    let fd = file.as_raw_fd();
+    asyncify(move || {
+        let fd = unsafe { BorrowedFd::borrow_raw(fd) };
+        rustix::fs::fstatvfs(fd).map_err(|e| e.into())
+    })
+    .await
+}
+
+async fn asyncify<F, T>(f: F) -> io::Result<T>
+where
+    F: FnOnce() -> io::Result<T> + Send + 'static,
+    T: Send + 'static,
+{
+    match spawn_blocking(f).await {
+        Ok(res) => res,
+        Err(_) => Err(io::Error::new(
+            io::ErrorKind::Other,
+            "background task failed",
+        )),
+    }
+}
+
+pub async fn dma_write_file_vectored<'a>(
+    path: impl AsRef<Path>,
+    bufs: &'a [IoSlice<'a>],
+) -> io::Result<usize> {
+    let mut file = DmaFile::create(path.as_ref()).await?;
+    let buf = DmaBuffer::new(file.alignment, file.alignment);
+    file.set_buffer(buf);
+
+    for buf in bufs {
+        let mut buf = &buf[..];
+
+        while !buf.is_empty() {
+            if file.buffer().remaining() == 0 {
+                file = asyncify(move || file.write_direct().map(|_| file)).await?;
+            }
+
+            let dst = file.mut_buffer();
+            let remaining = dst.remaining();
+            let n = buf.len().min(remaining);
+            let (left, right) = buf.split_at(n);
+            dst.extend_from_slice(left);
+            buf = right;
+        }
+    }
+
+    let file_length = bufs.iter().map(|buf| buf.len()).sum();
+    let len = file.buffer().len();
+    if len > 0 {
+        let align_up = file.align_up(len);
+        if align_up == len {
+            asyncify(move || file.write_direct().map(|_| file)).await?;
+        } else {
+            let dst = file.mut_buffer();
+            unsafe { dst.set_len(align_up) }
+            file = asyncify(move || file.write_direct().map(|_| file)).await?;
+            asyncify(move || file.truncate(file_length).map(|_| file)).await?;
+        }
+    }
+
+    Ok(file_length)
+}
+
+pub async fn dma_read_file(
```

Comment:
Maybe we can implement zero copy here.

Reply:
Do you mean `fn dma_read_file_all(path) -> buf` ?

Reply:
Just use the buffer passed to `dma_read_file directly`, instead of copying data to `DmaBuffer` first and then copying the data in `DmaBuffer` to the passed buffer.